devices/usbc/client.c

```diff
@@ -0,0 +1,167 @@
+/*
+ * Phoenix-RTOS
+ *
+ * phoenix-rtos-loader
+ *
+ * usb client
+ *
+ * Copyright 2019-2021 Phoenix Systems
+ * Author: Kamil Amanowicz, Hubert Buczynski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "../low.h"
+#include "../errors.h"
+
+#include "usbphy.h"
+#include "client.h"
+
+
+struct {
+	usb_dc_t dc;
+	usb_common_data_t data;
+} imx_common;
+
+
+int usbclient_send(int endpt, const void *data, unsigned int len)
+{
+	dtd_t *res;
+
+	if (len > USB_BUFFER_SIZE)
+		return -1;
+
+	if (!imx_common.data.endpts[endpt].caps[USB_ENDPT_DIR_IN].init)
+		return -1;
+
+	low_memcpy(imx_common.data.endpts[endpt].buf[USB_ENDPT_DIR_IN].buffer, data, len);
+	res = ctrl_execTransfer(endpt, (u32)imx_common.data.endpts[endpt].buf[USB_ENDPT_DIR_IN].buffer, len, USB_ENDPT_DIR_IN);
+
+	if (DTD_ERROR(res))
+		return -1;
+
+	return len - DTD_SIZE(res);
+}
+
+
+int usbclient_rcvEndp0(void *data, unsigned int len)
+{
+	int res = -1;
+
+	while ((imx_common.dc.op != DC_OP_RCV_ENDP0) && (imx_common.dc.op != DC_OP_RCV_ERR))
+
+	if (imx_common.dc.op != DC_OP_RCV_ERR) {
+		res = imx_common.data.endpts[0].buf[USB_ENDPT_DIR_OUT].len;
+		low_memcpy(data, (const char *)imx_common.data.endpts[0].buf[USB_ENDPT_DIR_OUT].buffer, res); /* copy data to buffer */
+		imx_common.dc.op = DC_OP_NONE;
+
+		ctrl_execTransfer(0, (u32)imx_common.data.endpts[0].buf[USB_ENDPT_DIR_IN].buffer, 0, USB_ENDPT_DIR_IN); /* ACK */
+	}
```

Comment:
Inner `if` condition is always `true`.

Reply:
I think that removing `(imx_common.dc.op != DC_OP_RCV_ERR)` from while loop is a better way. Then the 'if' condition makes sense.